crates/esthri/src/ops/download.rs

```diff
@@ -0,0 +1,234 @@
+            .map_ok(|(part, mut chunks)| async move {
+                let mut offset = (part - 1) * part_size;
+                while let Some(buf) = chunks.try_next().await? {
+                    let len = buf.len();
+                    write_all_at(Arc::clone(dest), buf, offset as u64).await?;
+                    offset += len as i64;
+                }
+                Result::Ok(())
+            })
+            .try_buffer_unordered(Config::global().concurrent_downloader_tasks());
+        let _: () = stream.try_collect().await?;
+    };
+
+    // If we're trying to download into a directory, assemble the path for the user
+    if download_path.is_dir() {
+        let s3filename = key
+            .split('/')
+            .next_back()
+            .ok_or(Error::CouldNotParseS3Filename)?;
+        dest.persist(download_path.join(s3filename)).await?;
+    } else {
+        dest.persist(download_path.into()).await?;
+    }
+
+    Ok(())
+}
+
+/// Fetches an object as a stream of `Byte`s
+fn download_streaming_helper<'a, T>(
+    s3: &'a T,
+    bucket: &'a str,
+    key: &'a str,
+    parts: u64,
+) -> impl Stream<Item = Result<Bytes>> + 'a
+where
+    T: S3,
+{
+    stream::iter(1..=parts)
+        .map(move |part| get_object_part_request(s3, bucket, key, part as i64))
+        .buffered(Config::global().concurrent_downloader_tasks())
+        .map_ok(GetObjectResponse::into_stream)
+        .try_flatten()
+}
+
+/// like download_streaming_helper but the requests are not in order and the part
+/// number is returned with each request
+fn download_unordered_streaming_helper<'a, T>(
+    s3: &'a T,
+    bucket: &'a str,
+    key: &'a str,
+    parts: u64,
+) -> impl Stream<Item = Result<(i64, impl Stream<Item = Result<Bytes>> + 'a)>> + 'a
+where
+    T: S3,
+{
+    stream::iter(1..=parts)
+        .map(move |part| get_object_part_request(s3, bucket, key, part as i64))
+        .buffer_unordered(Config::global().concurrent_downloader_tasks())
+        .map_ok(|res| (res.part, res.into_stream()))
+}
+
+async fn init_download_dir(path: &Path) -> Result<PathBuf> {
+    let mut path = path.to_owned();
+    tokio::task::spawn_blocking(move || {
+        if !path.is_file() && !path.is_dir() {
+            // If the specified destination directory doesn't already exist,
+            // see if the directory structure needs to be made
+            let dir = if !path.to_string_lossy().ends_with('/') {
+                path.pop();
+                path
+            } else {
+                path
+            };
+            if !dir.exists() && !dir.as_os_str().is_empty() {
+                info!("Creating directory path {:?}", dir.as_os_str());
+                std::fs::create_dir_all(&dir)?;
+            }
+            Ok(dir)
+        } else {
+            path.pop();
+            Ok(path)
+        }
+    })
+    .await?
+}
+
+#[cfg(unix)]
+async fn write_all_at(file: Arc<std::fs::File>, buf: Bytes, offset: u64) -> Result<()> {
```

Comment:
It would be interesting to add support for https://github.com/tokio-rs/tokio-uring as a feature. Should give a big boost to the write performance here